keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -1782,6 +1782,15 @@ Alpha:
 - Adequate documentation for the changes
 - Minimum viable test cases mentioned in Test Plan section
 
+Beta:
+
+- Supports concurrency limiting upon long-running requests
+- Debug endpoint dumping fine-grained states of the queues for priority-levels
+- Allow constant concurrency shares in the priority-level API model
+- Automatically manages versions of mandatory/suggested configuration
+- Necessary e2e test
+
```

Comment:
I would really like to have the LIST calls addressed though. Currently we treat "list all pods" as the same cost as "get single pod". I think this is much more important than addressing long-running requests.

Reply:
+1, and I think that's relatively easy to do.

Treat a LIST as weight = # of records requested. (unpaginated list, treat as a request for 10k items or something sufficiently punitive.)

Reply:
I'm not sure 10k is a good number, maybe we can introduce something a bit more dynamic here (we already expose #objects per type metric (it's exposed by periodically sending Count() request to etcd - IIRC once per minute). We can use those numbers for it and make it a bit more reflecting reality - the reason I don't fully like 10k is that it may be too small (by order of magnitude) in large clusters). But it's implementation details and we should discuss it outside of this doc :)

Reply:
I think the current plan is to focus on visibility & robustness for beta, and add features afterwards.

We can add features sooner--if we can find people to do the work.

Reply:
> Currently we treat "list all pods" as the same cost as "get single pod".

we've been discussing on discriminating unpaginated LIST calls from the very start --- we agreed it will be valuable for managing large clusters but didn't make it in alpha for simplicity. given the fact that unpaginated requests will eat-up/occupy more service-time in the apiserver, the current/alpha version is already applying some sort of discrimination upon unpaginated LIST calls linearly.. i think what @wojtek-t proposing is a significant/non-linear approach to punish those unproper LISTs in a way, (we can also punish those un-cached LISTs similarly).

Reply:
I think there's two pieces of work here:
1. The mechanism for acting on the estimate of an API call's cost
2. The rules that produce the cost estimation

We should make the code separate enough that it's easy for different people to work on these different tasks. It's easy to have someone go off and spend a month optimizing 2 once 1 is in place.

Reply:
Sure - (1) is much more important - we can work on tweaking (2) further on.